src/main/java/core/mate/academy/service/MachineServiceImpl.java

```diff
@@ -1,7 +1,39 @@
 package core.mate.academy.service;
 
-/**
- * Your implementation of MachineService.
- */
-public class MachineServiceImpl {
+import core.mate.academy.model.Bulldozer;
+import core.mate.academy.model.Excavator;
+import core.mate.academy.model.Machine;
+import core.mate.academy.model.Truck;
+import java.util.Collections;
+import java.util.List;
+
+public class MachineServiceImpl<T extends Machine> implements MachineService<T> {
+    @Override
+    public List<T> getAll(Class<? extends T> type) {
+        MachineProducer<? extends T> machineProducer;
+        if (type == Bulldozer.class) {
+            machineProducer = (MachineProducer<? extends T>) new BulldozerProducer();
```

Comment:
The cast `(MachineProducer<? extends T>)` is unchecked and could potentially lead to a `ClassCastException` at runtime if the type does not match. Consider using a safer approach to handle this casting.